core/shared/src/main/scala/org/typelevel/log4cats/SamLogger.scala

```diff
@@ -0,0 +1,154 @@
+
+  final def warn(logBit: LogRecord[Ctx], others: LogRecord[Ctx]*)(implicit
+      pkg: sourcecode.Pkg,
+      filename: sourcecode.FileName,
+      name: sourcecode.Name,
+      line: sourcecode.Line
+  ): F[Unit] = log_(KernelLogLevel.Warn, logBit, others: _*)
+
+  final def error(logBit: LogRecord[Ctx], others: LogRecord[Ctx]*)(implicit
+      pkg: sourcecode.Pkg,
+      filename: sourcecode.FileName,
+      name: sourcecode.Name,
+      line: sourcecode.Line
+  ): F[Unit] = log_(KernelLogLevel.Error, logBit, others: _*)
+
+  final def trace(logBit: LogRecord[Ctx], others: LogRecord[Ctx]*)(implicit
+      pkg: sourcecode.Pkg,
+      filename: sourcecode.FileName,
+      name: sourcecode.Name,
+      line: sourcecode.Line
+  ): F[Unit] = log_(KernelLogLevel.Trace, logBit, others: _*)
+
+  final def debug(logBit: LogRecord[Ctx], others: LogRecord[Ctx]*)(implicit
+      pkg: sourcecode.Pkg,
+      filename: sourcecode.FileName,
+      name: sourcecode.Name,
+      line: sourcecode.Line
+  ): F[Unit] = log_(KernelLogLevel.Debug, logBit, others: _*)
+
+  private final def log_(
+      level: KernelLogLevel,
+      bit: LogRecord[Ctx],
+      others: LogRecord[Ctx]*
+  )(implicit
+      pkg: sourcecode.Pkg,
+      filename: sourcecode.FileName,
+      name: sourcecode.Name,
+      line: sourcecode.Line
+  ): F[Unit] = {
+    log(
+      level,
+      (record: Builder) =>
+        LogRecord.combine[Ctx](others)(
+          bit(
+            record
+              .withLevel(level)
+              .withClassName(pkg.value + "." + name.value)
+              .withFileName(filename.value)
+              .withLine(line.value)
+          )
+        )
+    )
+  }
+
+  def withModifiedString(f: String => String): SamLogger[F, Ctx] =
+    SamLogger.withModifiedString[F, Ctx](this, f)
+  def mapK[G[_]](fk: F ~> G): SamLogger[G, Ctx] = SamLogger.mapK(fk)(this)
+}
+
+object SamLogger {
+  def apply[F[_], Ctx](implicit ev: SamLogger[F, Ctx]) = ev
+
+  def wrap[F[_], Ctx](kernel: LoggerKernel[F, Ctx]): SamLogger[F, Ctx] = new SamLogger[F, Ctx] {
+    def log(level: KernelLogLevel, record: Builder => Builder): F[Unit] =
+      kernel.log(level, record)
+  }
+
+  implicit def optionTSamLogger[F[_]: Functor, Ctx](implicit
+      ev: SamLogger[F, Ctx]
+  ): SamLogger[OptionT[F, *], Ctx] =
+    ev.mapK(OptionT.liftK[F])
+
+  implicit def eitherTSamLogger[F[_]: Functor, E, Ctx](implicit
+      ev: SamLogger[F, Ctx]
+  ): SamLogger[EitherT[F, E, *], Ctx] =
+    ev.mapK(EitherT.liftK[F, E])
+
+  implicit def kleisliSamLogger[F[_], A, Ctx](implicit
+      ev: SamLogger[F, Ctx]
+  ): SamLogger[Kleisli[F, A, *], Ctx] =
+    ev.mapK(Kleisli.liftK[F, A])
+
+  private def withModifiedString[F[_], Ctx](
+      l: SamLogger[F, Ctx],
+      f: String => String
+  ): SamLogger[F, Ctx] =
+    new SamLogger[F, Ctx] {
+      def log(level: KernelLogLevel, record: Builder => Builder): F[Unit] = {
+        val modifiedRecord = (builder: Builder) => {
```

Comment:
Needing to materialize the log and manually copy over the full message hints to me that we may need to refine `Builder`.

Pretty sure this would materialize `message` as well, so that's something to check.

Reply:
With the added `adaptMessage` method, this becomes:
```scala
      def log(level: KernelLogLevel, record: Builder => Builder): F[Unit] = {
        l.log(level, record(_).adaptMessage(f))
      }
```